### PR TITLE
[#172124428] Adds option to enable keepalive in fetch client

### PR DESCRIPTION
**Pull request by cloudify**

### CODE REVIEW

File: src/config.ts [hunk 1]
```diff
@@ -222,7 +227,42 @@ export const ALLOW_PAGOPA_IP_SOURCE_RANGE = decodeCIDRs(
...
+    return nodeFetch(input as any, initWithKeepalive as any) as any;
+  };
+};
+
+const apiFetch: typeof fetch = isFetchKeepaliveEnabled
+  ? getKeepaliveFetch()
+  : // tslint:disable-next-line: no-any
+    (nodeFetch as any);
+
+export const API_CLIENT = new ApiClientFactory(API_KEY, API_URL, apiFetch);
```
Comment (gunzip) [inline src/config.ts:265]: can you move this code into app.ts ? (this is a configuration file that should not have side effects)

File: src/config.ts [hunk 2]
```diff
@@ -222,7 +227,42 @@ export const ALLOW_PAGOPA_IP_SOURCE_RANGE = decodeCIDRs(
...
+// whether we want to reuse sockets in the fetch client
+const isFetchKeepaliveEnabled = process.env.FETCH_KEEPALIVE_ENABLED === "true";
+
+// Returns a fetch instance backed by a keepalive-enabled HTTP agent
+const getKeepaliveFetch: () => typeof fetch = () => {
+  // see https://nodejs.org/docs/latest-v10.x/api/http.html#http_new_agent_options
+  const keepAliveMaxSockets =
+    process.env.FETCH_KEEPALIVE_MAX_SOCKETS === undefined
+      ? Infinity
+      : parseInt(process.env.FETCH_KEEPALIVE_MAX_SOCKETS, 10);
```
Comment (gunzip) [inline src/config.ts:239]: I'd add this reference in a comment: https://blog.botframework.com/2018/03/05/fix-snat-exhaustion-node-js-bots/

see also this thread on using the library agentkeepalive (that has a ttl) vs native nodejs keepalive (which seems to work on newer version of node)

https://github.com/MicrosoftDocs/azure-docs/issues/29600#issuecomment-490354812

File: src/config.ts [hunk 3]
```diff
@@ -222,7 +227,42 @@ export const ALLOW_PAGOPA_IP_SOURCE_RANGE = decodeCIDRs(
...
 
-export const API_CLIENT = new ApiClientFactory(API_KEY, API_URL);
+// whether we want to reuse sockets in the fetch client
+const isFetchKeepaliveEnabled = process.env.FETCH_KEEPALIVE_ENABLED === "true";
+
+// Returns a fetch instance backed by a keepalive-enabled HTTP agent
+const getKeepaliveFetch: () => typeof fetch = () => {
+  // see https://nodejs.org/docs/latest-v10.x/api/http.html#http_new_agent_options
+  const keepAliveMaxSockets =
+    process.env.FETCH_KEEPALIVE_MAX_SOCKETS === undefined
```
Comment (gunzip) [inline src/config.ts:237]: this value must depend on nodejs processes, see

https://docs.microsoft.com/en-us/azure/app-service/app-service-web-nodejs-best-practices-and-troubleshoot-guide#my-node-application-is-making-excessive-outbound-calls

see also https://github.com/MicrosoftDocs/azure-docs/issues/8013

File: src/services/apiClientFactory.ts [hunk 4]
```diff
@@ -2,16 +2,28 @@
  * This service builds API clients.
  */
 
+import nodeFetch from "node-fetch";
+
 import { APIClient } from "../clients/api";
 import { IApiClientFactoryInterface } from "./IApiClientFactory";
 
+// TODO: this class is actually useless as APIClient is immutable, it can be removed
```
Comment (gunzip) [inline src/services/apiClientFactory.ts:10]: you can remove the class safely

File: src/config.ts (hunk 1 above)
Comment (cloudify) [reply src/config.ts:278]: that line was already there, I've just added a parameter, I tried to keep the changes at minimum

File: src/services/apiClientFactory.ts (hunk 4 above)
Comment (cloudify) [reply src/services/apiClientFactory.ts:10]: the class gets used in many places, it's better to do that in a separate PR

**Comment by gunzip** [comment]
@cloudify can you please provide some sensite default values so we can set them up into the backend app service ?

**Pull request merged**